### ISSUE TITLE
Stop auto focus on DropdownMenu.Item

### ISSUE THREAD

**Issue by itaikeren** [radix-ui/primitives#2193]
### Current Behavior

I want to put an input element as one of my dropdown menu items. 
My problem is that whenever I start to type something and move my cursor over other items, my input loses focus.

```
<DropdownMenu.Root>
    <DropdownMenu.Trigger />

    <DropdownMenu.Portal>
      <DropdownMenu.Content>
        <DropdownMenu.Item />
               <input ... />
          <DropdownMenu.Item />
        <DropdownMenu.Separator />
        <DropdownMenu.Item />
               text item
          <DropdownMenu.Item />
      </DropdownMenu.Content>
  </DropdownMenu.Root>

```

### Expected behavior

I want to be able to keep the focus on my input element (and keep typing) even if I move the cursor over other items. 

### Your environment

| Software         | Name(s) | Version |
| ---------------- | ------- | ------- |
| Radix Package(s) |    @radix-ui/dropdown-menu     |     2.0.4    |
| React            | n/a     |    18.2.0     |
| Node             | n/a     |    18.16.0     |
| npm/yarn         |         |         |
| Operating System |    macOS  Ventura    |    13.3.1     |

**Comment by vanlong2712** [comment]
Hi @itaikeren , any update?

**Comment by auduongtuan** [comment]
Hi, I have same issues, is there any update?

**Comment by wfortin** [comment]
Running into this issue. I'm using the Dropdown in a text format bar component (similar to Notion), and I want to preserve the focus on the selected text. It would be great to have some way of opting out of the autoFocus behavior.

**Comment by wmonecke** [comment]
Any workarounds? This is quite a pain.

**Comment by khanhtn91** [comment]
I have the same issue

**Comment by itaikeren** [comment]
Hey, sorry for the late response.
The only workaround I found for now is to add those lines to the `<Dropdown.Item>` component:

```js
<Dropdown.Item>
	...
	onPointerLeave={(event) => event.preventDefault()}
	onPointerMove={(event) => event.preventDefault()}
	...
</Dropdown.Item>
```

It will disable the auto-focus for each dropdown item when you hover over it.

**Comment by davidlky** [comment]
I got it working by wrapping my Input with a div, and adding a `onKeyDown={e=>e.stopPropagation()}`

like this
```javascript
<div className="sticky top-0 z-10 bg-white p-1" onKeyDown={(e) => e.stopPropagation()}>
  <Input
    placeholder={t.filter[lang]}
    value={search}
    onChange={(e) => {
      e.stopPropagation();
      e.preventDefault();
      setSearch(e.target.value);
      // bug in radix ui
      // setTimeout(() => e.target.focus(), 100);
    }}
  />
</div>
```

**Comment by nickzelei** [comment]
None of the above solutions worked for me. I tried overriding basically every mouse event I could find and nothing would stop the input box from losing focus when my mouse left.

**Comment by GraemeFulton** [comment]
> Running into this issue. I'm using the Dropdown in a text format bar component (similar to Notion), and I want to preserve the focus on the selected text. It would be great to have some way of opting out of the autoFocus behavior.

`onCloseAutoFocus` works for me: https://www.radix-ui.com/primitives/docs/components/dropdown-menu#content

```
<DropdownMenuContent
        onCloseAutoFocus={(e)=>{e.preventDefault(); e.stopPropagation();}}
...
```

**Comment by jamesallan93** [comment]
> I got it working by wrapping my Input with a div, and adding a `onKeyDown={e=>e.stopPropagation()}`
> 
> like this
> 
> ```js
> <div className="sticky top-0 z-10 bg-white p-1" onKeyDown={(e) => e.stopPropagation()}>
>   <Input
>     placeholder={t.filter[lang]}
>     value={search}
>     onChange={(e) => {
>       e.stopPropagation();
>       e.preventDefault();
>       setSearch(e.target.value);
>       // bug in radix ui
>       // setTimeout(() => e.target.focus(), 100);
>     }}
>   />
> </div>
> ```

this solution also works for shadcn-vue that uses  "radix-vue" as components

**Comment by kigary** [comment]
> None of the above solutions worked for me. I tried overriding basically every mouse event I could find and nothing would stop the input box from losing focus when my mouse left.

I have similar issue with Select component. Inside SelectContent I have input which is filtering selection. And when I type a letter which captures one option, that input is being blurred (the option is being focused). Also have a button which opens a dialog with a form. Tried everything, went through radix and shadcn docs and couldn't find a solution for disabling autofocus of an option. Ended up writing a bad workaround to focus back that input when it is being blurred if some conditions meet.
Something like this:
```
onFocus={() => {
  shouldRefocus = true;
}}
onBlur={(e) => {
  if (!shouldRefocus) return;
  e.currentTarget.focus();
}}
```

initially `shouldFocus` is `true`, on some event it is being set to `false`

**Comment by jamesallan93** [comment]
creating a wrapper with onClick to stopPropagation didn't work?, also with input onChange  e.stopPropagation();?

**Comment by kigary** [comment]
sorry for not providing whole scenario. Inside select I am using Command from [cmdk](https://cmdk.paco.me/), which is also enabling features like autofilter, pointer selection and looping over options. adding stopPropagation on input change causes filtering break. I was able to change to filter on my own, but ended up using blur/focus workaround to not implement what has already been implemented

**Comment by smzelek** [comment]
For future visitors:

My case was that I had a <Dialog> that launched after clicking an "edit" button which was in a <DropdownMenuItem>.
The Dialog contained a react-hook-form controlled input set to validate `onTouched` (focus + blur). Mysteriously, the input was immediately considered to be "touched" as soon as I clicked "edit" and the Dialog opened.

Turns out that while the DropdownMenuItem was animating out and the Dialog was animating in, the dying DropdownMenuItem was constantly trying to steal focus while the Dialog was trying to correctly set the new focus.

When I added logs, this meant that the main Input inside the Dialog's form was seeing this behavior:
```txt
> form rendered
> input focused
> input blurred
> input focused
> input blurred
> input focused
```

All happening in an instant as soon as the form rendered. (Before anyone asks, no, this had nothing to do with React Strict Mode).

Took a ton of debugging to find out that the DropdownMenuItem was the culprit, so this is the fix for future travelers to this issue:

```tsx
<DropdownMenuContent
  onCloseAutoFocus={(e) => editingFormIsOpen && e.preventDefault()}
>
  <DropdownMenuItem
    onPointerLeave={(e) => editingFormIsOpen && e.preventDefault()}
    onPointerMove={(e) => editingFormIsOpen && e.preventDefault()}
    onClick={() => openEditingForm(...)}
  >
    <Pencil />
    <span> Edit </span>
  </DropdownMenuItem>
</DropdownMenuContent>
```

When Edit is clicked and the edit form gets rendered, immediately start suppressing the problematic focus-stealing events on the Dropdown components. Using this conditional allows the Dropdown to behave properly in all other cases.

Cheers!

**Comment by trusktr** [comment]
This is indeed a pain.

I don't see exactly how this is a duplicate of #1969. This issue is far more clear with regards to the problem and need people have.


It would be nice if these automatic focusing behaviors were not applied by default. Or at least an option would be highly useful.

```jsx
return <DropdownMenu autoFocusItems={false} ...>
  ...
</DropdownMenu>
```

**Comment by MGSimard** [comment]
Same annoyance here. Rendering an autofocused input on the page upon clicking an + Add button.

- If I do it through a regular standalone button, the input autofocuses fine.
- If the add button is in a dropdownmenu, the input never autofocuses regardless of whether or not I use `onCloseAutoFocus`

**Comment by faridulreza** [comment]
none of the above worked for me. No `e.stopPropagation()` at any level helped. I also tried to manually get focus to the input after each change.

I tried this first:
```JS
const handleSearchChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchValue(e.target.value);
    if (searchInputRef.current) 
        searchInputRef.current.focus();
};
```
But it does not work due to timing issue. You see the dropdown item captures back the focus just after we force the focus.

So this dubious bad code works:
```JS
const handleSearchChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchValue(e.target.value);
    setTimeout(() => {
      if (searchInputRef.current) {
        searchInputRef.current.focus();
      }
    }, 1); // Ensure focus is set after state update
};
```

**Comment by DavidMetcalfe** [comment]
Hey — I came across this issue while investigating a similar focus-stealing bug in our own codebase, and I think the closure as a duplicate of #1969 may have been a mis-categorization. The two issues describe different behaviors:

- #2193 (this one) is about focus theft on `pointerMove` (mouse hover) inside `MenuItemImpl`.
- #1969 is about `aria-selected` roving focus during keyboard navigation.

Different code paths, different root causes, different fixes. I checked the latest `packages/react/menu/src/menu.tsx` and `MenuItemImpl.onPointerMove` still calls `item.focus({preventScroll: true})` unconditionally on every mouseover, with no opt-out. The 17+ users who piled workarounds onto this thread (`onPointerMove={e => e.preventDefault()}`, `onBlur` refocus, focus-timeout hacks, etc.) are all working around the same missing opt-out, and none of those workarounds are upstream.

I've drafted a focused fix at #3979 that adds a `focusOnItemEnter?: boolean` opt-out prop to `Menu.Content`, defaulting to `true` so existing users see zero behavior change. Setting it to `false` lets the consumer opt out of hover-to-focus when the menu contains focusable siblings (e.g. an embedded search `<input>`).

The key technical detail that distinguishes this from the user workarounds above: the prop gates only the `item.focus()` call, not the underlying `onItemEnter` event handling — so submenu hover-open via `MenuSubTrigger` (which uses a `setTimeout(open, 100)` on pointer-grace intent, not on focus) keeps working.

Could this be re-opened so the fix can be tracked here? Happy to revise the API shape if maintainers prefer a different name (`disableHoverFocus`, `autoFocusItems={false}` — both have been requested in this thread by other users). The diff is small (one behavior line + plumbing) and includes tests, a Storybook story, and a changeset.
